crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,404 @@
+                                int no_self_signed, int no_duplicates)
+{
+    int i;
+
+    if (sk == NULL)
+        return 0;
+
+    if (certs == NULL)
+        return 1;
+    for (i = 0; i < sk_X509_num(certs); i++) {
+        X509 *cert = sk_X509_value(certs, i);
+
+        if (!no_self_signed || X509_check_issued(cert, cert) != X509_V_OK) {
+            if (!OSSL_CMP_sk_X509_add1_cert(sk, cert, no_duplicates, 0))
+                return 0;
+        }
+    }
+    return 1;
+}
+
+int OSSL_CMP_X509_STORE_add1_certs(X509_STORE *store, STACK_OF(X509) *certs,
+                                   int only_self_signed)
+{
+    int i;
+
+    if (store == NULL)
+        return 0;
+
+    if (certs == NULL)
+        return 1;
+    for (i = 0; i < sk_X509_num(certs); i++) {
+        X509 *cert = sk_X509_value(certs, i);
+
+        if (!only_self_signed || X509_check_issued(cert, cert) == X509_V_OK)
+            if (!X509_STORE_add_cert(store, cert)) /* ups cert ref counter */
+                return 0;
+    }
+    return 1;
+}
+
+STACK_OF(X509) *OSSL_CMP_X509_STORE_get1_certs(X509_STORE *store)
+{
+    int i;
+    STACK_OF(X509) *sk;
+    STACK_OF(X509_OBJECT) *objs;
+
+    if (store == NULL)
+        return NULL;
+    if ((sk = sk_X509_new_null()) == NULL)
+        return NULL;
+    objs = X509_STORE_get0_objects(store);
+    for (i = 0; i < sk_X509_OBJECT_num(objs); i++) {
+        X509 *cert = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(objs, i));
+        if (cert != NULL) {
+            if (!sk_X509_push(sk, cert)) {
+                sk_X509_pop_free(sk, X509_free);
+                return NULL;
+            }
+            X509_up_ref(cert);
+        }
+    }
+    return sk;
+}
+
+/*-
+ * Builds up the certificate chain of certs as high up as possible using
+ * the given list of certs containing all possible intermediate certificates and
+ * optionally the (possible) trust anchor(s). See also ssl_add_cert_chain().
+ *
+ * Intended use of this function is to find all the certificates above the trust
+ * anchor needed to verify an EE's own certificate.  Those are supposed to be
+ * included in the ExtraCerts field of every first sent message of a transaction
+ * when MSG_SIG_ALG is utilized.
+ *
+ * NOTE: This allocates a stack and increments the reference count of each cert,
+ * so when not needed any more the stack and all its elements should be freed.
+ * NOTE: in case there is more than one possibility for the chain,
+ * OpenSSL seems to take the first one, check X509_verify_cert() for details.
+ *
+ * returns a pointer to a stack of (up_ref'ed) X509 certificates containing:
+ *      - the EE certificate given in the function arguments (cert)
+ *      - all intermediate certificates up the chain toward the trust anchor
+ *      - the (self-signed) trust anchor is not included
+ *      returns NULL on error
+ */
+STACK_OF(X509) *OSSL_CMP_build_cert_chain(STACK_OF(X509) *certs,
```

Comment:
Again why is this part of the public API? It doesn't feel CMP specific.

Reply:
Yes, see my comment above on all these utility functions.